enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/RaftServerModule.java

```diff
@@ -0,0 +1,117 @@
+        this.logProvider = platformModule.logging.getInternalLogProvider();
+    }
+
+    public void create( CoreServerModule coreServerModule )
+    {
+        LifecycleMessageHandler<ReceivedInstantClusterIdAwareMessage> messageHandlerChain = createMessageHandlerChain( coreServerModule );
+
+        createRaftServer( coreServerModule, messageHandlerChain );
+    }
+
+    private void createRaftServer( CoreServerModule coreServerModule, LifecycleMessageHandler<ReceivedInstantClusterIdAwareMessage> messageHandlerChain )
+    {
+        RaftServer raftServer = new RaftServer( new CoreReplicatedContentMarshal(), sslPolicy, platformModule.config, logProvider,
+                platformModule.logging.getUserLogProvider(), monitors, platformModule.clock );
+
+        LoggingInbound<ReceivedInstantClusterIdAwareMessage> loggingRaftInbound = new LoggingInbound<>( raftServer,
+                messageLogger, identityModule.myself() );
+        loggingRaftInbound.registerHandler( messageHandlerChain );
+
+        platformModule.life.add( raftServer ); // must start before core state so that it can trigger snapshot downloads when necessary
+        platformModule.life.add( coreServerModule.createCoreLife( messageHandlerChain ) );
+        platformModule.life.add( coreServerModule.catchupServer() ); // must start last and stop first, since it handles external requests
+    }
+
+    private LifecycleMessageHandler<ReceivedInstantClusterIdAwareMessage> createMessageHandlerChain( CoreServerModule coreServerModule )
+    {
+        RaftMessageApplier messageApplier = new RaftMessageApplier( localDatabase, logProvider,
+                consensusModule.raftMachine(), coreServerModule.downloadService(), coreServerModule.commandApplicationProcess() );
+
+        ComposableMessageHandler monitoringHandler = RaftMessageMonitoringHandler.composable( platformModule.clock, platformModule.monitors );
+
+        int queueSize = platformModule.config.get( CausalClusteringSettings.raft_in_queue_size );
+        int maxBatch = platformModule.config.get( CausalClusteringSettings.raft_in_queue_max_batch );
+        Function<Runnable, ContinuousJob> jobFactory = ( Runnable runnable ) ->
+                new ContinuousJob( platformModule.jobScheduler, new JobScheduler.Group( "raft-batch-handler", NEW_THREAD ), runnable, logProvider );
+        ComposableMessageHandler batchingMessageHandler = BatchingMessageHandler.composable( queueSize, maxBatch, jobFactory, logProvider );
+
+        ComposableMessageHandler leaderAvailabilityHandler =
+                LeaderAvailabilityHandler.composable( consensusModule.getLeaderAvailabilityTimers(), consensusModule.raftMachine()::term );
+
+        ComposableMessageHandler clusterBindingHandler = ClusterBindingHandler.composable( logProvider );
+
+        return clusterBindingHandler
+                .compose( leaderAvailabilityHandler )
+                .compose( batchingMessageHandler )
+                .compose( clusterBindingHandler )
```

Comment:
Should the cluster-binding really be twice in the stack here?

Reply:
It's the existing logic. I expect it's unnecessary.